sdk/python/lib/pulumi/provider/server.py

```diff
@@ -55,10 +54,20 @@ class ProviderServicer(ResourceProviderServicer):
     engine_address: str
     provider: Provider
     args: List[str]
+    lock: asyncio.Lock
+
 
-    # NOTE: remove @_asynchronized when we can avoid modifying globals in the method body.
-    @_asynchronized
```

Comment:
I was prematurely getting happy that we factored out the global state mod :) 

But looks like you just need to use the same lock object in both functions. In that case you could also just make it a parameter of the modifier so you say `@asynchronized(lambda self: self.lock)` and avoid the noise here. Doesn't matter too much just a tad more readable.

Reply:
I was originally planning to do that, but I'm hesitant because I'm not sure it'll use the same event loop as what will be used by the gRPC server.

Reply:
We could first move off threaded grpc server onto evented (asyncio) one? That'd remove threads.

Reply:
Or even in that case, there's a possibility of context switches in how we handle the request?